### ISSUE TITLE
Incorrect column names when using join on saved question with custom column

### ISSUE THREAD

**Issue by aurianepolleau8** [metabase/metabase#27494]
**Describe the bug**

When joining two saved question with one having custom columns, the names of the columns in the resulting table are incorrect. The names of the custom columns in joined question (right side of the join) become the name of the question on the left side of the join which is problematic because we cannot use those columns afterwards in a view, they get mixed up.  

When looking into the SQL for the question constructed with the query builder, the name of the columns is different for the source question and the joined question. 

**Logs** 

Not sure where to find them. 

**To Reproduce**

1. With the query builder, create a saved question with custom columns (here I used a case() to create a custom columns with the sample data). 

![image](https://user-images.githubusercontent.com/119337571/210524111-2d91b782-1fdf-47b4-b315-84198ef5b7fd.png)

3. Left join saved question to another saved question (simple aggregation) 

![image](https://user-images.githubusercontent.com/119337571/210524657-490f7d88-b9bd-40d8-80ef-c7def617c480.png)

 
4. We can see that the names of the column of the joined question become the same as the question on the left of the join
![image](https://user-images.githubusercontent.com/119337571/210525192-f6bdbd6b-7ffb-4838-a8a5-9be8c26bbbe4.png)

5. In the SELECT statement of the generated SQL, the columns are named as such: 
![image](https://user-images.githubusercontent.com/119337571/210525658-c9efc881-6cfc-45dc-921e-9ea52bc0ccec.png)

**Expected behavior**
I expect to either see the name of the column or something like sum_1, sum_2, sum_3 

**Information about your Metabase Installation:**

{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.16.1+1",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.16.1",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.16.1+1",
    "os.name": "Linux",
    "os.version": "5.4.209-116.367.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "bigquery-cloud-sdk"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.13"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.5.0"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2022-11-01",
      "tag": "v1.44.6",
      "branch": "release-x.44.x",
      "hash": "5c6ae21"
    },
    "settings": {
      "report-timezone": null
    }
  }
}

**Severity**
Low / Medium 
The joins are working fine but we cannot use the columns in views because the columns with similar names get mixed up / the users do not know what is in the columns anymore.

This means that I cannot let my non-sql users use the query builder to do these joins. I workaround that by creating mat views for these tables in the dwh but instead of leveraging metabase and let my users do it, I have to create a new view every time for these type of joins which is annoying. 

I could not find a similar bug report but please let me know of this is a known limitation or if there is another workaround I did not consider. Thanks!

**Comment by flamber** [comment]
Looks very related to #17769 and #12685/#12930/#21646

**Comment by uladzimirdev** [comment]
Indeed, column names were shown in a wrong way, but with the latest work on MLv2 it became better, please check the video attached. Column names are generated on BE side, even if there is still an issue, it's going to be a QPD scope

https://github.com/metabase/metabase/assets/125459446/1a76e505-5078-431b-bada-0503534843f2

**Comment by bshepherdson** [comment]
I agree with Uladzimir, the original _confusion_ of the columns is fixed.

But the presentation of the names is good in some places and bad in others. Column pickers seem to get this right, showing column names in context:
![2024-02-16-102921_1377x1017_scrot](https://github.com/metabase/metabase/assets/157812/61551c4a-2807-4273-ba8d-d22c0a0215e3)
![2024-02-16-102940_613x1043_scrot](https://github.com/metabase/metabase/assets/157812/1382cfa1-a3a7-4e4c-8348-7f6dd25e67f2)

while column headings showing the whole name in one string get it wrong:
![2024-02-16-103054_1520x905_scrot](https://github.com/metabase/metabase/assets/157812/615b0acc-9be1-46fd-9487-cc421e092606)

This seems like an issue with the MLv2-powered display names, and should be fixed. I'm updating the title of this issue and tagging it as an easy win for the MLv2 bug mountain.

**Comment by ranquild** [comment]
These column names come from QP and not MBQL lib

**Comment by ranquild** [comment]
This is fixed and already tested - column names from the QP are same as from the lib
